### PR TITLE
feat(auth): add support for Common Access Cards

### PR DESCRIPTION
**Pull request by eventualbuddha**
## Overview
VxSuite won't use `CommonAccessCard` directly, but `rave` will use it and is based on `vxsuite`. This is mostly about making rebases on `vxsuite/main` easier from within `rave`.

This includes two new scripts to aid in some common testing operations: checking a card PIN and getting a CAC certificate.

## Demo Video or Screenshot
```sh
❯ VX_MACHINE_TYPE=admin ./scripts/check-pin
Enter PIN: 000000
waitForReadyCardStatus: 1.001s
checkPin: 286.564ms
{ response: 'correct' }

❯ ./scripts/check-pin --cac
Enter PIN: 77777777
waitForReadyCardStatus: 0.112ms
checkPin: 419.776ms
{ response: 'correct' }
```

## Testing Plan
- [x] New automated tests
- [x] Tested VX Java cards manually with VxMark
- [x] Used the new `scripts/check-pin` script with both types of cards

### CODE REVIEW

File: libs/auth/src/common_access_card.ts [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+  }
+
+  async getCardStatus(): Promise<CardStatus<CommonAccessCardDetails>> {
+    return Promise.resolve(this.cardStatus);
+  }
+
+  async checkPin(pin: string): Promise<CheckPinResponse> {
+    await this.selectApplet();
+
+    const cardVxAdminCert = await this.getCertificate({
```
Comment (arsalansufi) [inline libs/auth/src/common_access_card.ts:157]: Probably want to rename this variable `cardDodCert`

**Review by arsalansufi — COMMENTED** [review]
Very cool to see how much we were able to reuse! My main comment is around moving all CAC code to a subdirectory

File: apps/admin/backend/src/server.ts [hunk 2]
```diff
@@ -73,8 +77,8 @@ export async function start({
       card:
         isFeatureFlagEnabled(BooleanEnvironmentVariableName.USE_MOCK_CARDS) ||
         isIntegrationTest()
-          ? new MockFileCard()
-          : new JavaCard(),
+          ? new MockFileVxSuiteCard()
+          : new VxSuiteJavaCard(),
```
[the remark this answers is not in the document]
Comment (eventualbuddha) [reply apps/admin/backend/src/server.ts:81]: Great, I like that. I'll relocate them into a `cac` directory and revert the `JavaCard` name change.

File: libs/auth/src/common_access_card.ts [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+    allFsPadding,
+    x00,
+    asn1Sha256MagicValue,
+    challengeHash,
+  ]);
+  assert(paddedMessage.byteLength === expectedPaddedMessageLength);
+
+  return new CardCommand({
+    ins: GENERAL_AUTHENTICATE.INS,
+    p1: 0x07, // TODO: what does this value mean?
```
[the remark this answers is not in the document]
Comment (eventualbuddha) [reply libs/auth/src/common_access_card.ts:86]: Great! Thanks for digging into this.

File: libs/auth/src/common_access_card.ts [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+    if (pin) {
+      await this.checkPinInternal(pin);
+    }
+
+    const generalAuthenticateResponse = await this.cardReader.transmit(
+      buildGenerateSignatureCardCommand(message, { privateKeyId })
+    );
+
+    // why are we trimming 8 here instead of 4, and why were we trimming 4 before? Is this an ECC vs. RSA formatting?
+    return generalAuthenticateResponse.subarray(8); // Trim metadata
```
[the remark this answers is not in the document]
Comment (eventualbuddha) [reply libs/auth/src/common_access_card.ts:253]: Thanks for all the info! After some more research I think I figured out what the first 8 bytes are. It looks like, for RSA 2048 or maybe just the CAC card, the first 8 bytes we're trimming are `7c 82 01 04 82 82 01 00`. This is a `GENERAL_AUTHENTICATE.DYNAMIC_AUTHENTICATION_TEMPLATE_TAG` object containing a `GENERAL_AUTHENTICATE.RESPONSE_TAG` object. So now I actually parse these with the expected tags so we don't try to return gibberish to openssl in the event the card responds with something weird.

[the remark this answers is not in the document]
Comment (eventualbuddha) [reply libs/auth/src/common_access_card.ts]: cc @benadida who originally wrote the trimming code

[the remark this answers is not in the document]
Comment (benadida) [reply libs/auth/src/common_access_card.ts]: I'm sorry I didn't add more comments to that and forced y'all to rediscover this!

[the remark this answers is not in the document]
Comment (arsalansufi) [reply libs/auth/src/common_access_card.ts]: No prob Ben, and thanks for fully confirming Brian!

**Review by arsalansufi — APPROVED** [review]
LGTM!

**Pull request merged**